---
title: "The URT microbiome of CRS patients: data analysis"
author: "Stijn Wittouck"
date: "November 20, 2019"
output:
  html_document:
    depth: 2
    number_sections: no
    theme: united
    toc: yes
    toc_float: yes
---

This Rmarkdown document describes the data analysis for the paper "Anterior nares diversity and pathobionts represent sinus microbiome in chronic rhinosinusitis" by De Boeck, Wittouck, Martens et al. All data visualizations and summary tables shown in the paper are generated by this document. You can read this document to get a more detailed overview of the data analysis, or you can run the code yourself and play around with it if you want to explore the data on your own. The easiest way to run the code is to use [RStudio](https://www.rstudio.com/products/rstudio/#Desktop). 

You need sevaral R packages to be able to run the code:

* tidyverse version 1.2.1
* ggrepel version 0.8.1 (to plot asv labels on scatterplots)
* ggpubr version 0.2 (to plot p-values on figures and compose figure panels)
* ggmosaic version 0.2.0 (for the mosaic plots)
* vegan version 2.5.4 (for the adonis tests)
* tidyamplicons version 0.2.0

You can install these packages by uncommenting and running the following commands:

```{r}
# install.packages("tidyverse")
# install.packages("ggrepel")
# install.packages("ggpubr")
# install.packages("ggmosaic")
# install.packages("vegan")
# install.packages("devtools") # required to install tidyamplicons from github
# devtools::install_github("SWittouck/tidyamplicons", ref = "v0.2.0")
```

Enjoy! 

# Set-up

Settings for rendering of this markdown file: 

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, message = F, warning = F)
```

Load the required R packages: 

```{r}
library(ape) # for plotting the clustering dendrogram
library(ggrepel)
library(ggpubr)
library(ggmosaic)
library(tidyamplicons)
library(tidyverse)
library(vegan)
```

Load some extra functions that we need (see functions.R for more information; the functions are documented in Roxygen style): 

```{r}
source("functions.R")
```

Make results folders: 

```{r}
outdirs <- c(
  "../results", "../results/locations", "../results/conditions", 
  "../results/polyps", "../results/metadata"
)
for (dir in outdirs) if (! dir.exists(dir)) dir.create(dir)
```

# Data preparation

We start by loading the data. All data is bundled in a "tidyamplicons" object: it contains the read counts of all taxa (ASVs) in all samples, the metadata of the samples (sampling location, patient characteristics, ...) and metadata of the taxa (taxonomic classification and interpretable taxon names). 

```{r}
load("../data/urt.rda")
```

Create a separate tidyamplicons object with abundances aggregated on the genus level: 

```{r}
urt_genus <- 
  urt %>%
  select_taxa(- species, - taxon_name, - sequence) %>%
  aggregate_taxa()
get_numbers(urt_genus)
```

Now we create a separate tidyamplicons object where the samples are rarefied to 1000 reads (samples with fewer reads will be dropped). At this point, we also set a seed to make the stochastic parts of this document reproducable.  

```{r}
set.seed(seed = 1991)
urt_r <- 
  urt %>%
  add_lib_size() %>%
  filter_samples(lib_size >= 1000) %>%
  tidyamplicons::rarefy(n = 1000, replace = F)
get_numbers(urt_r)
```

Define the order of the sampling locations for figures:

```{r}
locs <- c("anterior nares", "nasopharynx", "maxillary sinus", "ethmoid sinus")
```

# Exploration of metadata

Now that we prepared the dataset, we would like to explore the metadata of all CRS patients with at least one high-quality sample. The names of all metadata variables on the level of the participants (e.g. age), not samples (e.g. sampling location) is stored in a txt file called "variables.txt".

```{r}
predictors <- read_csv("../data/variables.txt")
crs_patients <- 
  urt$samples %>%
  filter(condition == "CRS") %>%
  select(participant, one_of(predictors$predictor), location) %>%
  group_by(participant) %>%
  mutate(n_samples = n()) %>%
  ungroup() %>%
  mutate(present = TRUE) %>%
  spread(key = location, value = present, fill = FALSE)
write_csv(crs_patients, "../results/table_S1_metadata.csv")
```

## Numerical variables (discrete and continuous)

Let's make a summary table of the numerical variables (both discrete and continuous ones). For each variable, we list the number of CRS patients that have data available, plus the mean and standard deviation of the variable: 

```{r}
av_table <- 
  crs_patients %>%
  select(!! predictors %>% filter(level != "cat") %>% pull(predictor)) %>%
  gather(value = "value", key = "variable", tidyselect::everything()) %>%
  group_by(variable) %>%
  summarize(
    not_na = value %>% na.omit %>% length,
    mean = mean(value, na.rm = T),
    sd = sd(value, na.rm = T)
  )
write_csv(av_table, "../results/metadata/predictors_summary_num.csv")
```

We can also explore the variables in a visualization: 

```{r}
crs_patients %>%
  select(!! predictors %>% filter(level == "disc") %>% pull(predictor)) %>%
  gather(value = "value", key = "variable", tidyselect::everything()) %>%
  ggplot(aes(x = value)) + 
  geom_bar() + 
  facet_wrap(~ variable, ncol = 1, scales = "free")
ggsave(
  "../results/metadata/expl_discrete.png", 
  units = "cm", width = 40, height = 20
)
crs_patients %>%
  select(!! predictors %>% filter(level == "cont") %>% pull(predictor)) %>%
  gather(value = "value", key = "variable", tidyselect::everything()) %>%
  ggplot(aes(x = value)) + 
  geom_density() + 
  facet_wrap(~ variable, ncol = 1, scales = "free")
ggsave(
  "../results/metadata/expl_continuous.png", 
  units = "cm", width = 40, height = 20
)
```

## Categorical variables

For the categorical variables we can make a frequency table: 

```{r}
freq_table <- 
  crs_patients %>%
  select(!! predictors %>% filter(level == "cat") %>% pull(predictor)) %>%
  gather(value = "value", key = "variable", tidyselect::everything()) %>%
  count(variable, value)
write_csv(freq_table, "../results/metadata/predictors_summary_cat.csv")
```

Visualization: 

```{r}
freq_table %>%
  ggplot(aes(x = value, y = n)) + 
  geom_col() + 
  coord_flip() + 
  facet_wrap(~ variable, scales = "free")
ggsave(
  "../results/metadata/expl_categorical.png", 
  units = "cm", width = 20, height = 15
)
```

## Sample dropout

Also interesting is the question whether the predictors are associated with the number of missing samples per participant? 

```{r}
crs_patients %>%
  select(
    !! predictors %>% filter(level %in% c("disc", "cont")) %>% pull(predictor),
    n_samples
  ) %>%
  gather(value = "value", key = "variable", - n_samples) %>%
  ggplot(aes(x = n_samples, y = value, group = n_samples)) + 
  geom_boxplot(outlier.alpha = 0) + 
  geom_jitter(width = 0.2, height = 0) + 
  facet_wrap(~ variable, ncol = 3, scales = "free")
ggsave(
  "../results/metadata/droupout_numeric.png", 
  units = "cm", width = 40, height = 20
)
```

Are the categorical predictors associated with the number of missing samples per participant? 

```{r}
crs_patients %>%
  select(
    !! predictors %>% filter(level %in% c("cat")) %>% pull(predictor),
    n_samples
  ) %>%
  gather(value = "value", key = "variable", - n_samples) %>%
  ggplot(aes(x = value, fill = as.factor(n_samples))) + 
  geom_bar(position = "dodge") + 
  facet_wrap(~ variable, ncol = 3, scales = "free") +
  scale_fill_brewer(palette = "Paired")
ggsave(
  "../results/metadata/droupout_categorical.png", 
  units = "cm", width = 40, height = 20
)
```

Is the dropout of samples from a particular sampling location associated with any of the predictors? 

```{r}
correlations <-
  crs_patients %>%
  select(- participant) %>%
  mutate(sex = if_else(sex == "M", 1, 0)) %>%
  cor(use = "pairwise.complete.obs") %>%
  as_tibble(rownames = "var_1") %>%
  gather(key = "var_2", value = "cor", - var_1) %>%
  mutate_at(
    c("var_1", "var_2"), factor, 
    levels = c(predictors$predictor, "SM", "SE", "N", "NF")
  ) %>%
  filter(as.numeric(var_1) < as.numeric(var_2))
correlations %>%
  ggplot(aes(x = cor)) +
  geom_density()
correlations %>%
  ggplot(aes(x = var_1, y = var_2, fill = cor)) +
  geom_tile() +
  theme(axis.text.x = element_text(angle = 90, hjust = 1))
correlations %>%
  arrange(desc(cor)) %>%
  print(n = 20)
```

# Results 1: comparison between niches

In the first results section, we compare the microbiomes of CRS patients between four niches (sampling locations): the nose (anterior nares), the nasopharynx, te maxillary sinus and the ethmoid sinus. 

## Top eleven genera (genus level)

In a first explorative plot, we visualize each sample as a vertical bar and give the eleven most abundant genera (overall) a color. In addition, we group the samples by niche:

```{r}
participants_ordered <- 
  urt_genus %>%
  select_samples(sample_id, participant) %>%
  aggregate_samples() %>%
  add_sample_clustered() %>%
  pluck("samples") %>%
  arrange(sample_clustered) %>%
  pull("participant")
(
  fig_locations_bar <- 
    urt_genus %>%
    filter_samples(condition == "CRS") %>%
    add_taxon_name_color(method = "total_rel_abundance") %>%
    modify_at("samples", mutate_at, "location", recode, 
      N = "anterior nares", 
      NF = "nasopharynx",
      SE = "ethmoid sinus",
      SM = "maxillary sinus"
    ) %>%
    modify_at("samples", mutate_at, "location", factor, levels = locs) %>%
    modify_at("samples", mutate_at, "participant", factor, levels = participants_ordered) %>%
    get_bar_plot(x = participant) +
    scale_fill_brewer(palette = "Paired", name = "genus") +
    facet_wrap(~ location, ncol = 1) +
    theme_bw() +
    theme(
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank(), 
      panel.grid = element_blank(),
      legend.position = "right",
      plot.margin = unit(c(0, 1, 0, 0), "cm")
    )
)
ggsave("../results/locations/barplot.png", units = "cm", width = 40, height = 20)
```

The genus name FM873692_g is not really interpretable; some research at the [EzBioCloud](https://www.ezbiocloud.net/) website learns us that it is a genus of the family Neisseriaceae.

We can also make a list of the overall most abundant genera in the CRS patients, across all niches: 

```{r}
abs_locations <- 
  urt_genus %>%
  filter_samples(condition == "CRS") %>%
  add_mean_rel_abundances(condition = "location") %>%
  taxa() %>%
  select(genus, contains("rel_abundance"))
abs_overall <-
  urt_genus %>%
  filter_samples(condition == "CRS") %>%
  add_mean_rel_abundances() %>%
  taxa() %>%
  select(genus, mean_rel_abundance)
left_join(abs_overall, abs_locations) %>%
  arrange(desc(mean_rel_abundance)) %>%
  slice(1:15) %>%
  write_csv("../results/table_S3_top_genera")
```

## Alpha diversity (asv level)

After our initial exploration, we can compare the four niches on their diversity, i.e. how many taxa samples from those niches contain on average. 

We will make a figure that shows the comparison between the four niches of two different alpha diversity measures: the ASV richness (simple count of the number of ASVs per sample) and the inverse Simpson measure. The latter takes "evenness" of the relative abundances into account; a sample will have a lower inverse Simpson when a few (or one) taxa are very dominent (or even one taxon in the extreme case). 

We start by making a table with the diversity values: 

```{r}
diversities_locations <- 
  urt_r %>%
  filter_samples(condition == "CRS") %>%
  filter_samples(location %in% c("N", "NF", "SM", "SE")) %>%
  add_diversity_measures() %>%
  pluck("samples") %>%
  mutate_at("location", recode, 
    N = "anterior nares", NF = "nasopharynx", SM = "maxillary sinus", SE = "ethmoid sinus"
  ) %>%
  mutate_at("location", factor, levels = locs) %>%
  select(participant, location, contains("div")) %>%
  gather(value = "diversity", key = "metric", contains("div"))%>%
  mutate_at("metric", recode,
    div_inv_simpson = "inverse Simpson", 
    div_observed = "richness"
  ) 
```

Now we do pairiwse t-tests between the sampling locations, both for the richness and for the inverse simpson values: 

```{r}
comparisons <-
  diversities_locations %>%
  group_by(metric) %>%
  do(pairwise_t_tests(., value = diversity, group = location)) %>%
  mutate(p_value = p.adjust(p_value, method = "holm")) %>%
  mutate(label = case_when(
    p_value < 0.0001 ~ "****",
    p_value < 0.001 ~ "***",
    p_value < 0.01 ~ "**",
    p_value < 0.05 ~ "*",
    TRUE ~ ""
  ))
print(comparisons, n = 20)
```

Now we can visualize these results: 

```{r}
fig_locations_div_raw <-
  diversities_locations %>%
  ggplot(aes(x = location, y = diversity)) +
  geom_boxplot(fill = "grey90", col = "grey60", outlier.alpha = 0) +
  geom_jitter(width = 0.1, height = 0, alpha = 1, size = 0.1) +
  facet_wrap(~ metric, scales = "free_y", ncol = 1) +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.5))
(
  fig_locations_div <-
    fig_locations_div_raw %>%
    add_comparisons(
      comparisons %>% filter(p_value < 0.05), 
      group_levels = locs, facet_vars = "metric", vscale = 0.10,
      size_brackets = 0.2, size_labels = 2
    )
)
ggsave("../results/locations/diversity.png", units = "cm", width = 17.4, height = 20)
```

Remark: we have used the rarefied data for this; otherwise a sample with many reads would tend to seem more diverse. It might be good to show that samples with large read depth do not have higher inverse simpson values: 

```{r}
urt %>%
  add_lib_size() %>%
  add_alphas() %>%
  pluck("samples") %>%
  ggplot(aes(x = lib_size, y = inverse_simpson)) + 
  geom_point() +
  scale_x_log10() +
  scale_y_log10() +
  theme_bw()
ggsave(
  "../results/locations/inv_simpson_read_depth.png", 
  units = "cm", width = 17.4, height = 17.4
)
```

## Beta diversity (asv level)

Pairwise beta diversity is a measure for how similar (or different) two samples are. We use the Bray-Curtis dissimilarity (because of rather technical reasons, we should not call it a distance), which can be interpreted as the percentage of the contents of one sample that is not present in the other one. 

If we try to make a two-dimensional scatterplot of the samples where the distances between the points are as close as possible (relatively speaking) to the Bray-Curtis distances, we are making a principal coordinates analysis (PCoA) plot: 

```{r}
urt %>%
  filter_samples(condition == "CRS") %>%
  add_pcoa() %>%
  {pcoa_vars <<- .$pcoa_variances; .} %>%
  pluck("samples") %>%
  ggplot(aes(x = pcoa1, y = pcoa2, col = location)) +
  geom_point() +
  scale_color_brewer(palette = "Paired") +
  xlab(paste0("PCoA component 1 (", round(pcoa_vars[1] * 100), "%)")) +
  ylab(paste0("PCoA component 2 (", round(pcoa_vars[2] * 100), "%)")) + 
  theme_bw()
ggsave("../results/locations/pcoa.png", units = "cm", width = 17.4, height = 12)
ggsave("../results/figure_S1_locations_pcoa.tiff", units = "cm", width = 17.4, height = 12)
```

The PCoA plot is an approximation in two dimensions; a lot of information gets lost. Now we will visualize all pairwise Bray-Curtis distances. Calculating those distances happened "under the hood" when we used the `add_pcoa()` function. Therefore, we have to calculate them again, this time to store them in a table: 

```{r}
crs <- filter_samples(urt, condition == "CRS")
betas <- 
  crs %>%
  get_betas(method = "bray", binary = F) %>%
  split(f = if_else(.$location_1 >= .$location_2, "yes", "no")) %>%
  modify_at("no", mutate, location_1_new = location_2, location_2_new = location_1) %>%
  modify_at("no", mutate, location_1 = location_1_new, location_2 = location_2_new) %>%
  modify_at("no", select, - location_1_new, - location_2_new) %>%
  do.call(what = bind_rows)
```

Now we visualize all those distances in boxplots, grouped per type of "comparison" that they represent: 

```{r}
locations_levels <- c(
  "N_N", "NF_NF", "SM_SM", "SE_SE",
  "NF_N", "SE_N", "SM_N", "SE_NF", "SM_NF", "SM_SE"
)

locations_labels <- c(
  "anterior nares",
  "nasopharynx",
  "maxillary sinus",
  "ethmoid sinus",
  "nasopharynx - \nanterior nares",
  "ethmoid sinus - \nanterior nares",
  "maxillary sinus - \nanterior nares",
  "ethmoid sinus - \nnasopharynx",
  "maxillary sinus - \nnasopharynx",
  "maxillary sinus - \nethmoid sinus"
)

(
  fig_locations_betas <- 
    betas %>%
    mutate(locations = str_c(location_1, location_2, sep = "_")) %>%
    mutate_at("locations", factor, levels = locations_levels, labels = locations_labels) %>%
    mutate(same_participant = if_else(participant_1 == participant_2,
      "within-subject", "between-subject"
    )) %>%
    mutate(same_location = if_else(location_1 == location_2,
      "within-location", "between-location"
    )) %>%
    mutate_at("same_location", factor, levels = c("within-location", "between-location")) %>% 
    ggplot(aes(x = locations, y = beta, col = locations, fill = locations)) +
    geom_violin(draw_quantiles = 0.5, alpha = 0.3) +
    facet_grid(~ same_participant + same_location, scales = "free_x", space = "free_x") +
    scale_color_brewer(palette = "Paired") +
    scale_fill_brewer(palette = "Paired") + 
    ylab("bray-curtis dissimilarity") +
    theme_bw() +
    theme(
      axis.text.x = element_blank(),
      axis.ticks.x = element_blank(),
      legend.text = element_text(margin = margin(b = 5))
    )
)
ggsave("../results/locations/betas.png", units = "cm", width = 20, height = 16)
```

Finally, we perform an adonis test with the variables location (niche) and participant als predictors. This will tell us if the microbiome differs significantly between the locations (within participants) and/or between the participants (within locations). 

```{r}
perform_adonis(crs, predictors = c("location", "participant"))
```

## Panel

We arrange the figures of results section 1 into one large panel: 

```{r}
resize <- function() {
  m <- 0.1
  theme(
    text = element_text(size = 7),
    legend.key.size = unit(0.8, "line"),
    strip.text = element_text(margin = margin(m, m, m, m, "cm"))
  )
}
ggarrange(
  (fig_locations_bar + resize()) %>% give_letter("A"), 
  ggarrange(
    (fig_locations_div + resize()) %>% give_letter("B"), 
    (fig_locations_betas + resize()) %>% give_letter("C"),
    ncol = 2, widths = c(3, 5)
  ),
  nrow = 2,
  heights = c(1, 1)
)
ggsave("../results/figure_1_locations.tiff", units = "cm", width = 17.5, height = 17.5, dpi = 300)
```

# Results 2: comparison between conditions

In the second results section, we compare the microbiome of the CRS patients to that of ~ 100 healthy volunteers. 

## Overall adonis

Adonis with "condition" (CRS/CON) as prediction, on the nose samples only:

```{r}
urt %>%
  filter_samples(location == "N") %>%
  perform_adonis(predictors = "condition")
```

Same test, this time on the nasopharynx samples: 

```{r}
urt %>%
  filter_samples(location == "NF") %>%
  perform_adonis(predictors = "condition")
```

## Alpha diversity (asv level)

Here we compare alpha diversity (richness and inverse Simpson) between: healthy, CRS with polyps and CRS without polyps. We exclude CRS patients with polyps NA. 

We start by making a table with the diversity values: 

```{r}
diversities_conditions <-
  urt_r %>%
  filter_samples(location %in% c("N", "NF")) %>%
  add_diversity_measures() %>%
  pluck("samples") %>%
  mutate_at("location", recode, 
    N = "anterior nares", NF = "nasopharynx"
  ) %>%
  mutate_at("location", factor, levels = locs) %>%
  mutate(case = case_when(
    condition == "CON" ~ "healthy",
    polyps == 0 ~ "CRSsNP",
    polyps == 1 ~ "CRSwNP",
    is.na(polyps) ~ "crs (polyps NA)"
  )) %>%
  filter(case != "crs (polyps NA)") %>%
  mutate(case = factor(case, levels = c("healthy", "CRSsNP", "CRSwNP"))) %>%
  select(case, location, contains("div")) %>%
  gather(value = "diversity", key = "metric", contains("div")) %>%
  mutate_at(
    "metric", recode,
    div_inv_simpson = "inverse Simpson", 
    div_observed = "richness"
  )
```

Now we do pairiwse t-tests between the three conditions, both for the richness and for the inverse simpson values and for the nose and nasopharynx: 

```{r}
comparisons <-
  diversities_conditions %>%
  group_by(metric, location) %>%
  do(pairwise_t_tests(., value = diversity, group = case)) %>%
  mutate(p_value = p.adjust(p_value, method = "holm")) %>%
  mutate(label = case_when(
    p_value < 0.0001 ~ "****",
    p_value < 0.001 ~ "***",
    p_value < 0.01 ~ "**",
    p_value < 0.05 ~ "*",
    TRUE ~ ""
  ))
print(comparisons, n = 20)
```

Now we can visualize these results: 

```{r}
fig_conditions_div_raw <-
  diversities_conditions %>%
  ggplot(aes(x = case, y = diversity)) +
  geom_boxplot(fill = "grey90", col = "grey60", outlier.alpha = 0) +
  geom_jitter(width = 0.1, height = 0, alpha = 1, size = 1) +
  facet_grid(metric ~ location, scales = "free_y") +
  theme_bw() +
  theme(axis.text.x = element_text(angle = 45, vjust = 0.5))
fig_conditions_div_raw %>%
  add_comparisons(
    comparisons %>% filter(p_value < 0.05), 
    group_levels = c("healthy", "CRSsNP", "CRSwNP"), 
    facet_vars = c("metric", "location")
  )
ggsave("../results/conditions/alpha_diversity.tiff", units = "cm", width = 17.4, height = 16)
ggsave("../results/figure_2_conditions_diversity.tiff", units = "cm", width = 17.4, height = 16)
```

## Differential presence of CON vs CRS (asv level)

In this part, we compare the occurrence of ASVs between the CRS and CON subjects. More specifically, we look for ASVs that are significantly more present in one of the two subject groups than in the other. 

We create a separate presence/absence table for nose and nasopharynx samples and we merge them: 

```{r}
pres_nose <- 
  urt_r %>%
  filter_samples(location == "N") %>%
  add_occurrences(condition = "condition", relative = T, fischer_test = T) %>%
  taxa() %>%
  mutate(location = "anterior nares")
pres_npx <- 
  urt_r %>%
  filter_samples(location == "NF") %>%
  add_occurrences(condition = "condition", relative = T, fischer_test = T) %>%
  taxa() %>%
  mutate(location = "nasopharynx")
pres_table <- 
  bind_rows(pres_nose, pres_npx) %>%
  mutate(
    fisher_sign = if_else(fisher_p < 0.05, "p < 0.05", "p >= 0.05"),
    taxon_name_plot = if_else(
      (occurrence_in_CON > 0.25 | occurrence_in_CRS > 0.25) & fisher_p < 0.05,
      taxon_name, as.character(NA)
    ),
    location = factor(location, levels = c("anterior nares", "nasopharynx"))
  )
```

Now we can visualize the "occurrences" and fischer exact test significances in a scatterplot: 

```{r}
(
  fig_conditions_pres <- 
    pres_table %>%
    ggplot(aes(
      x = occurrence_in_CON, y = occurrence_in_CRS, col = fisher_sign
    )) +
    geom_point() +
    stat_function(fun = function(x) x, col = "grey") +
    geom_label_repel(aes(label = taxon_name_plot)) +
    facet_wrap(~ location, nrow = 1) + 
    xlim(c(0, 1)) + ylim(c(0, 1)) +
    scale_color_manual(
      values = c("p < 0.05" = "#1f78b4", "p >= 0.05" = "#a6cee3"), 
      name = "fisher exact test"
    ) +
    xlab("presence in CON samples") +
    ylab("presence in CRS samples") +
    theme_bw()
)
ggsave(
  "../results/conditions/differential_presence.png", 
  units = "cm", width = 35, height = 15
)
```

## Differential abundance of CON vs CRS (asv level)

This part is similar to the previous part, but here we determine the mean relative abundances of all taxa in the CON and CRS groups, and whether those differ significantly (using a wilcoxon test). 

```{r}
ab_nose <- 
  urt %>%
  filter_samples(location == "N") %>%
  add_mean_rel_abundances("condition", test = "t-test") %>%
  pluck("taxa") %>%
  mutate(location = "anterior nares")
ab_npx <- 
  urt %>%
  filter_samples(location == "NF") %>%
  add_mean_rel_abundances("condition", test = "t-test") %>%
  pluck("taxa") %>%
  mutate(location = "nasopharynx")
ab_table <- 
  bind_rows(ab_nose, ab_npx) %>%
  mutate(
    location = factor(location, levels = c("anterior nares", "nasopharynx")),
    t_test_sign = if_else(t_test_p < 0.05, "p < 0.05", "p >= 0.05"),
    taxon_name_plot = if_else(
      (mean_rel_abundance_in_CON > 0.03 | mean_rel_abundance_in_CRS > 0.03), 
      taxon_name, as.character(NA)
    ) 
  )
```

Visualize differential abundances in a scatterplot: 

```{r}
(
  fig_conditions_abun <- 
    ab_table %>%
    ggplot(aes(
      x = mean_rel_abundance_in_CON, y = mean_rel_abundance_in_CRS, 
      col = t_test_sign
    )) +
    geom_point() +
    stat_function(fun = function(x) x, col = "grey") +
    geom_label_repel(aes(label = taxon_name_plot)) +
    facet_wrap(~ location, nrow = 1) +
    xlim(c(0, 0.20)) + ylim(c(0, 0.20)) +
    scale_color_manual(
      values = c("p < 0.05" = "#1f78b4", "p >= 0.05" = "#a6cee3"), 
      name = "welch t-test"
    ) +
    xlab("mean relative abundance in CON samples") +
    ylab("mean relative abundance in CRS samples") +
    theme_bw() 
)
ggsave(
  "../results/conditions/differential_abundance.png", 
  units = "cm", width = 35, height = 15
)
```

We can now summarize the ASVs that were significantly differentially present and/or abundant in a table: 

```{r}
bind_rows(
  pres_table %>%
    filter(fisher_p < 0.05) %>%
    transmute(
      taxon_name, taxon_name_plot,
      preference = if_else(
        occurrence_in_CON > occurrence_in_CRS, "healthy", "CRS"
      ),
      location,
      comparison = "occurrence"
    ),
  ab_table %>%
    filter(t_test_p < 0.05) %>%
    transmute(
      taxon_name, taxon_name_plot,
      preference = if_else(
        mean_rel_abundance_in_CON > mean_rel_abundance_in_CRS, "healthy", "CRS"
      ),
      location,
      comparison = "abundance"
    )
) %>%
  transmute(
    taxon_name, taxon_name_plot, preference, 
    header = str_c(comparison, location, sep = " in ")
  ) %>%
  group_by(taxon_name) %>%
  mutate(keep_taxon = sum(! is.na(taxon_name_plot)) > 0) %>%
  ungroup() %>%
  filter(keep_taxon) %>%
  select(taxon_name, header, preference) %>%
  spread(key = header, value = preference) %>%
  left_join(urt$taxa %>% select(taxon_name, species)) %>%
  write_csv("../results/table_S2_differential_taxa.csv")
```

## Differential presence of CRSsNP vs CRSwNP

Now we test differential presence for CRSsNP vs CRSwNP: 

```{r}
pres_nose <- 
  urt_r %>%
  filter_samples(! is.na(polyps), location == "N") %>%
  modify_at(
    "samples", mutate_at, "polyps", ~ if_else(. == 1, "CRSwNP", "CRSsNP")
  ) %>%
  add_occurrences(condition = "polyps", relative = T, fischer_test = T) %>%
  taxa() %>%
  mutate(location = "anterior nares")
pres_npx <- 
  urt_r %>%
  filter_samples(! is.na(polyps), location == "NF") %>%
  modify_at(
    "samples", mutate_at, "polyps", ~ if_else(. == 1, "CRSwNP", "CRSsNP")
  ) %>%
  add_occurrences(condition = "polyps", relative = T, fischer_test = T) %>%
  taxa() %>%
  mutate(location = "nasopharynx")
pres_table <- 
  bind_rows(pres_nose, pres_npx) %>%
  mutate(
    fisher_sign = if_else(fisher_p < 0.05, "p < 0.05", "p >= 0.05"),
    taxon_name_plot = if_else(
      (occurrence_in_CRSsNP > 0.25 | occurrence_in_CRSwNP > 0.25) & 
        fisher_p < 0.05,
      taxon_name, as.character(NA)
    ),
    location = factor(location, levels = c("anterior nares", "nasopharynx"))
  )
(
  fig_polyps_pres <- 
    pres_table %>%
    ggplot(aes(
      x = occurrence_in_CRSsNP, y = occurrence_in_CRSwNP, col = fisher_sign
    )) +
    geom_point() +
    stat_function(fun = function(x) x, col = "grey") +
    geom_label_repel(aes(label = taxon_name_plot)) +
    facet_wrap(~ location, nrow = 1) + 
    xlim(c(0, 1)) + ylim(c(0, 1)) +
    scale_color_manual(
      values = c("p < 0.05" = "#1f78b4", "p >= 0.05" = "#a6cee3"), 
      name = "fisher exact test"
    ) +
    xlab("presence in CRSsNP samples") +
    ylab("presence in CRSwNP samples") +
    theme_bw() 
)
ggsave(
  "../results/polyps/differential_presence.png", 
  units = "cm", width = 35, height = 15
)
```

## Differential abundance of CRSsNP vs CRSwNP

And finally we test differential abundance for CRSsNP vs CRSwNP: 

```{r}
ab_nose <- 
  urt %>%
  filter_samples(! is.na(polyps), location == "N") %>%
  modify_at(
    "samples", mutate_at, "polyps", ~ if_else(. == 1, "CRSwNP", "CRSsNP")
  ) %>%
  add_mean_rel_abundances("polyps", test = "t-test") %>%
  pluck("taxa") %>%
  mutate(location = "anterior nares")
ab_npx <- 
  urt %>%
  filter_samples(! is.na(polyps), location == "NF") %>%
  modify_at(
    "samples", mutate_at, "polyps", ~ if_else(. == 1, "CRSwNP", "CRSsNP")
  ) %>%
  add_mean_rel_abundances("polyps", test = "t-test") %>%
  pluck("taxa") %>%
  mutate(location = "nasopharynx")
ab_table <- 
  bind_rows(ab_nose, ab_npx) %>%
  mutate(
    location = factor(location, levels = c("anterior nares", "nasopharynx")),
    t_test_sign = if_else(t_test_p < 0.05, "p < 0.05", "p >= 0.05"),
    taxon_name_plot = if_else(
      (mean_rel_abundance_in_CRSsNP > 0.03 | mean_rel_abundance_in_CRSwNP > 0.03),
      taxon_name, as.character(NA)
    ) 
  )
(
  fig_polyps_abun <- 
    ab_table %>%
    ggplot(aes(
      x = mean_rel_abundance_in_CRSsNP, y = mean_rel_abundance_in_CRSwNP, 
      col = t_test_sign
    )) +
    geom_point() +
    stat_function(fun = function(x) x, col = "grey") +
    geom_label_repel(
      aes(label = taxon_name_plot)
    ) +
    facet_wrap(~ location, nrow = 1) +
    xlim(c(0, 0.22)) + ylim(c(0, 0.22)) +
    scale_color_manual(
      values = c("p < 0.05" = "#1f78b4", "p >= 0.05" = "#a6cee3"), 
      name = "welch t-test"
    ) +
    xlab("mean relative abundance in CRSsNP samples") +
    ylab("mean relative abundance in CRSwNP samples") +
    theme_bw() 
)
ggsave(
  "../results/polyps/differential_abundance.png", 
  units = "cm", width = 35, height = 15
)
```

## Nose-nasopharynx continuity 

Is the continuity between the nose and nasopharynx that we observed in the CRS patients different in the healthy controls? 

```{r}
betas <- 
  urt %>%
  filter_samples(location %in% c("N", "NF")) %>%
  get_betas(method = "bray", binary = F) %>%
  filter(participant_1 == participant_2) %>%
  filter(location_1 != location_2)
betas %>%
  ggplot(aes(x = condition_1, y = beta)) +
  geom_violin(draw_quantiles = 0.5) +
  stat_compare_means(comparisons = list(c(1, 2)), method = "t.test") +
  xlab("condition") +
  ylab("within-subject Bray-Curtis dissimilarity")
ggsave(
  "../results/conditions/continuity.png", 
  units = "cm", width = 12, height = 10
)
```

Apparently not. 

## Microbiome variance of CON vs CRS 

Another possible extra research question is whether the microbiome of CRS patients is more diverse (between participants) than that of healthy controls. 

```{r}
betas_nose_con <-
  urt %>%
  filter_samples(location == "N", condition == "CON") %>%
  betas()
betas_nose_crs <-
  urt %>%
  filter_samples(location == "N", condition == "CRS") %>%
  betas() 
betas_npx_con <-
  urt %>%
  filter_samples(location == "NF", condition == "CON") %>%
  betas()
betas_npx_crs <-
  urt %>%
  filter_samples(location == "NF", condition == "CRS") %>%
  betas() 
betas_nose_con %>%
  bind_rows(betas_nose_crs) %>%
  bind_rows(betas_npx_con) %>%
  bind_rows(betas_npx_crs) %>%
  ggplot(aes(x = condition_1, y = beta)) +
  geom_violin(draw_quantiles = 0.5, alpha = 0.3) + 
  facet_wrap(~ location_1) +
  ylab("bray-curtis dissimilarity") +
  xlab("condition") +
  theme_bw()
ggsave(
  "../results/conditions/beta_diversity.png", units = "cm", width = 20, 
  height = 16
)
```

## Panels

We put the differential occurrences and differential abundances of CON vs CRS together in one figure panel:

```{r}
ggarrange(
  fig_conditions_pres %>% give_letter("A"), 
  fig_conditions_abun %>% give_letter("B"),
  ncol = 1, nrow = 2
)
ggsave(
  "../results/figure_3_conditions_diffpresabun.tiff", 
  units = "cm", width = 30, height = 25
)
```

Same for CRSsNP vs CRSwNP: 

```{r}
ggarrange(
  fig_polyps_pres %>% give_letter("A"), 
  fig_polyps_abun %>% give_letter("B"),
  ncol = 1, nrow = 2
)
ggsave(
  "../results/figure_S2_polyps_diffpresabun.tiff", 
  units = "cm", width = 30, height = 25
)
```

# Results 3: assocations with metadata

In this final results section, we check for associations between the microbiome of the CRS patients and various types of metadata: patient characteristics (such as age), phenotypes (such as allergies) and endotypes (such as IL-5 blood levels). We do this only for the nasopharynx samples. 

## Associations with metadata (asv level)

We make a separate tidyamplicons object with only the nasopharynx samples of CRS patients: 

```{r}
crs_nf <- filter_samples(urt, condition == "CRS", location == "NF")
```

One very important variable that divides the CRS patients into two subgroups is whether they have polyps. We first check if this variable is associated with the microbiome using an adonis test: 

```{r}
perform_adonis(crs_nf, "polyps")
```

Now we perform an adonis test for all predictors, separately on the CRSsNP and CRSwNP groups: 

```{r}
predictors_enriched <- 
  predictors %>%
  filter(predictor != "polyps") %>%
  mutate(CRSsNP = map(predictor, ~ perform_adonis(
    crs_nf %>% filter_samples(polyps == 0), .
  ))) %>%
  mutate(CRSwNP = map(predictor, ~ perform_adonis(
    crs_nf %>% filter_samples(polyps == 1), .
  ))) %>%
  mutate(all_subjects = map(predictor, ~ perform_adonis(
    crs_nf, .
  ))) %>%
  gather(key = "polyps", value = "adonis", CRSsNP, CRSwNP, all_subjects) %>%
  mutate(R2 = map_dbl(adonis, list("aov.tab", "R2", 1))) %>%
  mutate(p = map_dbl(adonis, list("aov.tab", "Pr(>F)", 1))) %>%
  mutate(n = map_int(adonis, ~ nrow(.[["model.matrix"]]))) %>%
  select(- adonis) %>%
  mutate(sign = p < 0.05)
predictors_enriched %>%
  write_csv("../results/metadata/predictors_pvalues.csv")
```

Visualize results:

```{r}
predictors_enriched %>%
  arrange(polyps, R2) %>%
  mutate(predictor = factor(predictor, levels = unique(predictor))) %>%
  ggplot(aes(x = predictor, y = R2, fill = type, shape = sign)) +
  geom_col() +
  geom_point(size = 5) + 
  geom_text(aes(label = n, y = - 0.001)) + 
  facet_wrap(~ polyps) +
  scale_fill_brewer(
    palette = "Paired",
    guide = guide_legend(override.aes = list(shape = NA), title = "type of predictor")
  ) + 
  scale_shape_manual(
    values = c(`TRUE` = 42, `FALSE` = 32),
    guide = guide_legend(title = "significant")
  ) +
  coord_flip() +
  theme_bw() 
ggsave("../results/metadata/predictors.png", units = "cm", width = 30, height = 10)
ggsave("../results/figure_4_predictors.tiff", units = "cm", width = 30, height = 10)
```

In our previous paper on the URT microbiome of healthy volunteers, we found a significant association between the microbiome and smoking. Let's visualize this variable (smoking) on a bar chart of the nasopharynx samples: 

```{r}
crs_nf %>%
  get_bar_plot() +
  geom_rug(data = crs_nf$samples, aes(x = sample_id, col = as.factor(smoking), fill = NULL, y = NULL)) +
  scale_color_brewer(palette = "Paired") +
  theme(axis.text.x = element_blank())
```

## Association of microbiome clusters with metadata (genus level)

In this subsection, we inspect the associations of the microbiome with the metadata in a bit more detail by dividing our CRS subject population in different "microbiome types" extracted from a hierarchical clustering. We extract fourteen clusters from the clustering tree and remove clusters with less than five samples. 

```{r}
crs_genus_clusters <- 
  urt_genus %>%
  filter_samples(condition == "CRS") %>%
  filter_samples(location == "NF") %>%
  add_cluster(n_clusters = 14) %>%
  modify_at("samples", add_count, cluster) %>%
  filter_samples(n >= 5)
crs_genus_clusters %>%
  get_bar_plot() +
  geom_point(aes(y = - 0.05, col = cluster)) +
  scale_color_brewer(palette = "Paired") +
  theme(
    axis.text.x = element_blank()
  )
ggsave("../results/metadata/clusters.png", units = "cm", width = 30, height = 20)
ggsave("../results/figure_S3_clusters.tiff", units = "cm", width = 30, height = 20)
```

We also plot a dendrogram of the hierarchical clustering: 

```{r}
clust <- 
  crs_genus_clusters %>%
  get_rel_abundance_matrix() %>%
  vegdist() %>%
  hclust(method = "average") %>% 
  as.phylo() %>%
  keep.tip(tip = crs_genus_clusters$samples$sample_id)
tiff(filename = "../results/figure_S3_clusters_dendro.tiff")
plot(clust, show.tip.label = F, direction = "downwards")
dev.off()
```

Now we can create a summary table, with the mean of all numerical predictors per "subject cluster". 

```{r}
crs_genus_clusters %>%
  samples() %>%
  select(
    cluster, 
    one_of(predictors %>% filter(level != "cat") %>% pull(predictor))
  ) %>%
  group_by(cluster) %>%
  summarize_all(mean, na.rm = T)
```

Exploratory visualization of the numerical predictors: 

```{r}
(
  fig_predictors_num <- 
    crs_genus_clusters %>%
    samples() %>%
    select(
      cluster, 
      one_of(predictors %>% filter(level != "cat") %>% pull(predictor))
    ) %>%
    gather(key = "predictor", value = "value", - cluster) %>%
    ggplot(aes(x = cluster, y = value)) + 
    geom_boxplot(outlier.alpha = 0) + 
    geom_jitter(height = 0, width = 0.2, alpha = 1, col = "grey") + 
    facet_wrap(~ predictor, scales = "free_y", ncol = 2) +
    theme_bw()
)
ggsave("../results/metadata/predictors_num.png", units = "cm", width = 20, height = 15)
```

Exploratory visulizations of the categorical predictors: 

```{r}
predictors_tidy <- 
  crs_genus_clusters %>%
  samples() %>%
  select(
    cluster, 
    one_of(predictors %>% filter(level == "cat") %>% pull(predictor))
  ) %>%
  mutate(male = sex == "M") %>%
  select(- sex) %>%
  gather(key = "predictor", value = "value", - cluster) %>%
  mutate_at("value", as.character) %>%
  filter(! is.na(value))

predictors_tidy %>%
  ggplot(aes(x = cluster, fill = value)) + 
  geom_bar(position = "stack") + 
  facet_wrap(~ predictor, scales = "free", ncol = 2) +
  scale_fill_brewer(palette = "Paired") + 
  theme_bw()
ggsave("../results/metadata/predictors_cat_1.png", units = "cm", width = 20, height = 15)

predictors_tidy %>%
  ggplot(aes(x = value, y = cluster)) +
  geom_count(aes(col = ..n..)) +
  facet_wrap(~ predictor) + 
  theme_bw()
ggsave("../results/metadata/predictors_cat_2.png", units = "cm", width = 20, height = 15)

(
  fig_predictors_cat <-
    predictors_tidy %>%
    mutate(value = if_else(value == 1, "yes", "no")) %>%
    ggplot() +
    geom_mosaic(aes(x = product(value, cluster), fill = value)) +
    coord_flip() + 
    facet_wrap(~ predictor) +
    scale_fill_brewer(palette = "Paired") + 
    theme_bw() +
    theme(
      axis.title = element_blank(),
      axis.ticks.x = element_blank(),
      axis.text.x = element_blank(),
      panel.grid = element_blank()
    )
)
ggsave("../results/metadata/predictors_cat_3.png", units = "cm", width = 20, height = 15)

predictors_tidy %>%
  mutate(value = if_else(value == 1, "yes", "no")) %>%
  ggplot() +
  geom_mosaic(aes(x = product(cluster, value), fill = cluster)) +
  coord_flip() + 
  facet_wrap(~ predictor) +
  scale_fill_brewer(palette = "Paired") + 
  theme_bw() +
  theme(
    axis.title = element_blank(),
    axis.ticks.x = element_blank(),
    axis.text.x = element_blank(),
    panel.grid = element_blank()
  )
ggsave("../results/metadata/predictors_cat_4.png", units = "cm", width = 20, height = 15)
```

## Panel

Again, we can put the exploratory visualizatons of the predictors vs the subject clusters in a figure panel: 

```{r}
ggarrange(
  fig_predictors_num %>% give_letter("A"), 
  fig_predictors_cat %>% give_letter("B"), 
  ncol = 1, nrow = 2
)
ggsave("../results/figure_5_predictors_clusters.tiff", units = "cm", width = 35, height = 30)
```

